lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -3065,8 +3066,49 @@ bool LLVMToSPIRVBase::transDecoration(Value *V, SPIRVValue *BV) {
           }
         }
       }
-      if (M != 0)
+      // Handle nofpclass attribute. Nothing to do if fast math flag is already
+      // set.
+      if ((BV->isExtInst() &&
+           static_cast<SPIRVExtInst *>(BV)->getExtSetKind() ==
+               SPIRVEIS_OpenCL) &&
+          BM->isAllowedToUseVersion(VersionNumber::SPIRV_1_6) &&
+          !(M & FPFastMathModeFastMask)) {
+        auto *F = cast<CallInst>(V)->getCalledFunction();
+        auto FAttrs = F->getAttributes();
+        AttributeSet RetAttrs = FAttrs.getRetAttrs();
+        if (RetAttrs.hasAttribute(Attribute::NoFPClass)) {
+          FPClassTest RetTest =
+              RetAttrs.getAttribute(Attribute::NoFPClass).getNoFPClass();
+          AttributeSet RetAttrs = FAttrs.getRetAttrs();
+          // Only Nan and Inf tests are representable in SPIR-V now.
+          bool ToAddNoNan = RetTest & fcNan;
+          bool ToAddNoInf = RetTest & fcInf;
+          if (ToAddNoNan || ToAddNoInf) {
+            const size_t NumParams = F->getFunctionType()->getNumParams();
+            for (size_t I = 0; I != NumParams; ++I) {
+              if (!F->hasParamAttribute(I, Attribute::NoFPClass)) {
```

Comment:
There are extended instructions with non-FP parameters (e.g. ldexp).

Reply:
Yes, there are builtins, where one can not apply the Decoration, above there is a comment from me:

> Ideally we should also check, if fast math flags are applicable to a particular function from OpenCL std lib according to environmental spec, but I beg your pardon for not doing that.

I can't imagine something better, then to check the names of the builtins in this case, and it will be not officiant and TBH might be even unnecessary (see below).

> non-FP parameters

AFAIU clang or any other LLVM IR producer won't put fast math flag and / or attribute in this case as LLVM wouldn't allow it, there are checks if an LLVM value is castable to FPMathOperator resulting in errors/assertions. So I'm not really expecting such IR be ever an input to the translator.

Reply:
The logic of your code is "use FP Fast Math flag is **all** parameters" use the same `nofpclass` attribute". 
I suggest we change it to "use FP Fast Math flag is **all FP** parameters" use the same `nofpclass` attribute".

Reply:
Got it, so the suggestion is to:
if a function is `float nofpclass(nan) foo(float nofpclass(nan), i32, i32)` - we still need to apply the `NotNaN` FP math mode.

Reply:
Right. 👍